go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,467 @@
+				subdir := filepath.Base(filepath.Dir(path))
+				k, err := getKeyFromFileName(subdir + d.Name())
+				if err != nil {
+					return fmt.Errorf("error parsing cached file name %s: %w", path, err)
+				}
+				info, err := d.Info()
+				if err != nil {
+					return fmt.Errorf("error getting file info of %s: %w", path, err)
+				}
+				it := &qitem{
+					key: k,
+					lat: fileInfoToAccessTime(info),
+				}
+				size, err := res.getItemSize(k)
+				if err != nil {
+					return fmt.Errorf("error getting file size of %s: %w", path, err)
+				}
+				res.store.Store(k, it)
+				atomic.AddInt64(&res.stats.TotalSizeBytes, size)
+				atomic.AddInt64(&res.stats.TotalNumFiles, 1)
+				res.mu.Lock()
+				heap.Push(res.queue, it)
+				res.mu.Unlock()
+				return nil
+			})
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	go res.gc()
+	return res, nil
+}
+
+func (d *DiskCache) getItemSize(k key) (int64, error) {
+	if k.isCas {
+		return k.digest.Size, nil
+	}
+	fname := d.getPath(k)
+	info, err := os.Stat(fname)
+	if err != nil {
+		return 0, fmt.Errorf("error getting info for %s: %w", fname, err)
+	}
+	return info.Size(), nil
+}
+
+// Terminates the GC daemon, waiting for it to complete. No further Store* calls to the DiskCache should be made.
+func (d *DiskCache) Shutdown() {
+	d.shutdownOnce.Do(func() {
+		d.shutdown <- true
+		<-d.gcDone
+		log.Infof("DiskCacheStats: %+v", d.stats)
+	})
+}
+
+func (d *DiskCache) GetStats() *DiskCacheStats {
+	// Return a copy for safety.
+	return &DiskCacheStats{
+		TotalSizeBytes:         atomic.LoadInt64(&d.stats.TotalSizeBytes),
+		TotalNumFiles:          atomic.LoadInt64(&d.stats.TotalNumFiles),
+		NumFilesStored:         atomic.LoadInt64(&d.stats.NumFilesStored),
+		TotalStoredBytes:       atomic.LoadInt64(&d.stats.TotalStoredBytes),
+		NumFilesGCed:           atomic.LoadInt64(&d.stats.NumFilesGCed),
+		TotalGCedSizeBytes:     atomic.LoadInt64(&d.stats.TotalGCedSizeBytes),
+		NumCacheHits:           atomic.LoadInt64(&d.stats.NumCacheHits),
+		NumCacheMisses:         atomic.LoadInt64(&d.stats.NumCacheMisses),
+		TotalCacheHitSizeBytes: atomic.LoadInt64(&d.stats.TotalCacheHitSizeBytes),
+		InitTime:               time.Duration(atomic.LoadInt64((*int64)(&d.stats.InitTime))),
+		TotalGCTime:            time.Duration(atomic.LoadInt64((*int64)(&d.stats.TotalGCTime))),
+	}
+}
+
+func getKeyFromFileName(fname string) (key, error) {
+	pair := strings.Split(fname, ".")
+	if len(pair) != 2 {
+		return key{}, fmt.Errorf("expected file name in the form hash[_ac].size, got %s", fname)
+	}
+	size, err := strconv.ParseInt(pair[1], 10, 64)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid size in digest %s: %s", fname, err)
+	}
+	hash, isAc := strings.CutSuffix(pair[0], "_ac")
+	dg, err := digest.New(hash, size)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid digest from file name %s: %w", fname, err)
+	}
+	return key{digest: dg, isCas: !isAc}, nil
+}
+
+func (d *DiskCache) getPath(k key) string {
```

Comment:
Optional nit: I would have preferred files to have a uniform file name "%s-%d" % (hash, size), instead of having the size as an extension. Then you could use ".cas" as the marker for cas files. IMHO it would be a bit cleaner.